moveit_ros/occupancy_map_monitor/test/occupancy_map_monitor_tests.cpp

```diff
@@ -0,0 +1,81 @@
+ *     with the distribution.
+ *   * Neither the name of PickNik nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Tyler Weaver */
+
+#include <memory>
+#include <string>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "moveit/occupancy_map_monitor/occupancy_map_monitor.h"
+#include "tf2_ros/buffer.h"
+
+struct MockMiddlewareHandle : public occupancy_map_monitor::OccupancyMapMonitor::MiddlewareHandle
+{
+  MOCK_METHOD(double, getMapResolutionParameter, (), (override));
+  MOCK_METHOD(std::string, getMapFrameParameter, (), (override));
+  MOCK_METHOD(std::string, getSensorPluginParameter, (const std::string& sensor), (override));
+  MOCK_METHOD(std::vector<std::string>, getSensorListParameter, (), (override));
+  MOCK_METHOD(occupancy_map_monitor::OccupancyMapUpdaterPtr, loadOccupancyMapUpdater,
+              (const std::string& sensor_plugin), (override));
+  MOCK_METHOD(void, initializeOccupancyMapUpdater,
+              (occupancy_map_monitor::OccupancyMapUpdaterPtr occupancy_map_updater), (override));
+  MOCK_METHOD(void, createSaveMapService, (occupancy_map_monitor::SaveMapServiceCallback callback), (override));
+  MOCK_METHOD(void, createLoadMapService, (occupancy_map_monitor::LoadMapServiceCallback callback), (override));
+};
+
+TEST(OccupancyMapMonitorTests, ConstructorTest)
+{
+  // GIVEN a mocked middlware handle
+  auto mock_middleware_handle = std::make_unique<MockMiddlewareHandle>();
+
+  // THEN we expect it to call these methods on the MiddlewareHandle
+  EXPECT_CALL(*mock_middleware_handle, getMapFrameParameter);
+  EXPECT_CALL(*mock_middleware_handle, getSensorListParameter);
+  EXPECT_CALL(*mock_middleware_handle, getMapResolutionParameter);
+  EXPECT_CALL(*mock_middleware_handle, createSaveMapService);
+  EXPECT_CALL(*mock_middleware_handle, createLoadMapService);
+
+  // WHEN we start the current state monitor
+  occupancy_map_monitor::OccupancyMapMonitor occupancy_map_monitor{
```

Comment:
Isn't it GIVEN, WHEN, THEN? Or am I missing something?

Reply:
Yes, but because of how gmock works it goes in this order.  The expect macro has to come before the creation of the OccupancyMapMonitor where the mock object is moved into it.